packages/app/src/systems/DApp/machines/connectRequestMachine.ts

```diff
@@ -183,26 +193,39 @@ export const connectRequestMachine = createMachine(
           return undefined;
         },
       }),
-      addConnection: FetchMachine.create<Connection, Connection | undefined>({
+      addConnection: FetchMachine.create<
+        MachineContext,
+        Connection | undefined
+      >({
         showError: false,
         fetch: async ({ input }) => {
-          if (!input?.origin || !Array.isArray(input?.accounts)) {
-            throw new Error('Origin or account not passed');
+          if (
+            !input?.origin ||
+            !Array.isArray(input?.selectedAddresses) ||
+            input?.selectedAddresses.length === 0
+          ) {
+            throw new Error('Invalid connect request');
+          }
+          if (input.connection) {
+            return ConnectionService.updateConnectedAccounts({
+              origin: input.origin,
+              accounts: input.selectedAddresses,
+            });
           }
           return ConnectionService.addConnection({
             data: {
               origin: input.origin,
-              accounts: input.accounts,
-              favIconUrl: input.favIconUrl,
-              title: input.title,
+              accounts: input.selectedAddresses,
+              favIconUrl: input.favIconUrl!,
+              title: input.title!,
             },
           });
         },
       }),
     },
     guards: {
-      connectionConnected: (_, ev) => {
-        return !!((ev.data?.accounts.length || 0) >= 1);
+      connectionConnected: (ctx, ev) => {
+        return (ev.data?.accounts.length || 0) === ctx.totalAccounts;
```

Comment:
@luizstacio should we also add a check here that `ctrx.totalAccounts >= 1`?

Reply:
yeah 
```suggestion
        return (ev.data?.accounts.length || 0) === ctx.totalAccounts && ctx.totalAccounts >= 1
```